lib/matplotlib/text.py

```diff
@@ -66,15 +66,26 @@ def _get_text_metrics_with_cache(renderer, text, fontprop, ismath, dpi):
     """Call ``renderer.get_text_width_height_descent``, caching the results."""
     # Cached based on a copy of fontprop so that later in-place mutations of
     # the passed-in argument do not mess up the cache.
-    return _get_text_metrics_with_cache_impl(
-        weakref.ref(renderer), text, fontprop.copy(), ismath, dpi)
-
-
-@functools.lru_cache(4096)
-def _get_text_metrics_with_cache_impl(
-        renderer_ref, text, fontprop, ismath, dpi):
-    # dpi is unused, but participates in cache invalidation (via the renderer).
-    return renderer_ref().get_text_width_height_descent(text, fontprop, ismath)
+    func = _get_text_metrics_function(renderer)
+    return func(text, fontprop.copy(), ismath, dpi)
+
+
+# use mutable default to carry hidden global state
+def _get_text_metrics_function(inp_renderer, _cache=weakref.WeakKeyDictionary()):
+    if (_text_metrics := _cache.get(inp_renderer, None)) is None:
+        renderer_ref = weakref.ref(inp_renderer)
+        @functools.lru_cache(4096)
+        def _text_metrics(text, fontprop, ismath, dpi):
+            # dpi is unused, but participates in cache invalidation (via the renderer).
+            if (lcl_renderer := renderer_ref()) is None:
```

Comment:
And I assume `lcl` here is `local` as well? Probably can fit the two letters here too...